modules/partition-distribution/src/main/java/org/apache/ignite/internal/partitiondistribution/AssignmentsQueue.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.partitiondistribution;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.Objects;
+import org.apache.ignite.internal.tostring.IgniteToStringInclude;
+import org.apache.ignite.internal.tostring.S;
+import org.apache.ignite.internal.versioned.VersionedSerialization;
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Class that encapsulates a queue with consequent configuration switches
+ * e.g. promoting a learner to a peer or demoting a peer to a learner in following steps:
+ * <ul>
+ *   <li> node removed from configuration
+ *   <li> node added to configuration with different type (peer or learner)
+ * </ul>
+ */
+public class AssignmentsQueue implements Iterable<Assignments> {
+
+    @IgniteToStringInclude
+    private final Deque<Assignments> queue;
+
+    /** Constructor. */
+    public AssignmentsQueue(Assignments... assignments) {
+        this(Arrays.asList(assignments));
+    }
+
+    /** Constructor. */
+    AssignmentsQueue(Collection<Assignments> assignments) {
+        this.queue = new LinkedList<>(assignments);
+    }
+
+    /**
+     * Retrieves and removes the head of this queue, or returns {@code Assignments.EMPTY} if this queue is empty.
+     *
+     * @return the head of this queue, or {@code Assignments.EMPTY} if this queue is empty
+     */
+    public @Nullable Assignments poll() {
+        return queue.poll();
+    }
```

Comment:
Javadoc is now incorrect, because the method is nullable. Why don't you want to add an assertion that the queue is not empty? It should never return null